lib/fluent/plugin/out_sumologic.rb

```diff
@@ -409,46 +415,57 @@ def write(chunk)
       start_time = Time.now
       sleep_time = @retry_min_interval
 
-      common_log_part = "#{@data_type} records with source category '#{source_category}', source host '#{source_host}', source name '#{source_name}', chunk #{chunk_id}, try #{retries}"
-      while true
-        begin
-          @log.debug { "Sending #{messages.count}; #{common_log_part}" }
-
-          @sumo_conn.publish(
-              messages.join("\n"),
-              source_host         =source_host,
-              source_category     =source_category,
-              source_name         =source_name,
-              data_type           =@data_type,
-              metric_data_format  =@metric_data_format,
-              collected_fields    =fields,
-              dimensions          =@custom_dimensions
-          )
-          break
-        rescue => e
-          if @skip_retry
-            raise e
-          end
-          # increment retries
-          retries = retries + 1
-
-          log.warn "error while sending request to sumo: #{e}; #{common_log_part}"
-          log.warn_backtrace e.backtrace
+      if @records_per_request <= 0
+        iterations = 0
+      else
+        iterations = (messages.length/@records_per_request.to_f).ceil -1
+      end
 
-          # drop data if
-          #   - we reached out the @retry_max_times retries
-          #   - or we exceeded @retry_timeout
-          if (retries >= @retry_max_times && @retry_max_times > 0) || (Time.now > start_time + @retry_timeout && @retry_timeout > 0)
-            log.warn "dropping records; #{common_log_part}"
+      for i in 0..iterations do
+        # this will be 0..-1 for @records_per_request equal to 0
+        messages_to_send = messages[i*@records_per_request..(i+1)*@records_per_request-1]
```

Comment:
This is too magical for me when `records_per_request` and `iterations` are both 0. Can we instead split `messages` into chunks and iterate over the chunks here? Then we can have special logic for `records_per_request=0` and won't have to rely on these subtle loop and slice semantics.